armsrc/em4x50.c

```diff
@@ -0,0 +1,877 @@
+        [14] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [15] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [17] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [18] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [19] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [20] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [21] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [22] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [23] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [24] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [25] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [26] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [27] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [28] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [29] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [30] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [31] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // user
+        [32] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // device serial number
+        [33] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, // device identification
+    },
+};
+
+// Sam7s has several timers, we will use the source TIMER_CLOCK1 (aka AT91C_TC_CLKS_TIMER_DIV1_CLOCK)
+// TIMER_CLOCK1 = MCK/2, MCK is running at 48 MHz, Timer is running at 48/2 = 24 MHz
+// EM4x50 units (T0) have duration of 8 microseconds (us), which is 1/125000 per second (carrier)
+// T0 = TIMER_CLOCK1 / 125000 = 192
+
+#ifndef T0
+#define T0                                  192
+#endif
+
+#define EM4X50_T_TAG_QUARTER_PERIOD         16
+#define EM4X50_T_TAG_HALF_PERIOD            32
+#define EM4X50_T_TAG_THREE_QUARTER_PERIOD   48
+#define EM4X50_T_TAG_FULL_PERIOD            64
+#define EM4X50_T_WAITING_FOR_LIW            500
+#define EM4X50_T_TAG_TPP                    64
+#define EM4X50_T_TAG_TWA                    64
+
+#define EM4X50_TAG_TOLERANCE                8
+#define EM4X50_TAG_WORD                     45
+
+#define EM4X50_BIT_0                        0
+#define EM4X50_BIT_1                        1
+#define EM4X50_BIT_OTHER                    2
+
+#define EM4X50_COMMAND_LOGIN                0x01
+#define EM4X50_COMMAND_RESET                0x80
+#define EM4X50_COMMAND_WRITE                0x12
+#define EM4X50_COMMAND_WRITE_PASSWORD       0x11
+#define EM4X50_COMMAND_SELECTIVE_READ       0x0A
+
+#define FPGA_TIMER_0                        0
+
+// auxiliary functions
+
+static void init_tag(void) {
+
+    // initialize global tag structure
+    
+    for (int i = 0; i < 34; i++)
+        for (int j = 0; j < 7; j++)
+            tag.sectors[i][j] = 0x00;
+}
+
+static uint8_t bits2byte(uint8_t bits[8], int length) {
```

Comment:
send ref?

Reply:
What do you mean?

Reply:
bits[8]  ..  use  uint8_t *bits   instead

Reply:
done